src/Materials/effectRenderer.ts

```diff
@@ -0,0 +1,126 @@
+                if (!this._tmpScreenBuffer) {
+                    var internalTexture = this.engine.createRenderTargetTexture(
+                        {
+                            width: Math.floor(this.engine.getRenderWidth(true)),
+                            height: Math.floor(this.engine.getRenderHeight(true)),
+                        },
+                        {
+                            generateDepthBuffer: false,
+                            generateStencilBuffer: false,
+                            generateMipMaps: false,
+                            samplingMode: Constants.TEXTURE_NEAREST_NEAREST,
+                        },
+                    );
+                    this._tmpScreenBuffer = new Texture("", null);
+                    this._tmpScreenBuffer._texture = internalTexture;
+                }
+                renderTo = this._tmpScreenBuffer;
+            }else {
+                renderTo = outputTexture;
+            }
+
+            // for any next effect make it's input the output of the previous effect
+            if (i !== 0) {
+                effectWrapper.effect.onBindObservable.addOnce(() => {
+                    effectWrapper.effect.setTexture("textureSampler", this._tmpScreenBuffer);
+                });
+            }
+
+            // Reset state
+            this.engine.setViewport(new Viewport(0, 0, 1, 1));
+            this.engine.enableEffect(effectWrapper.effect);
+
+            // Bind buffers
+            if (renderTo) {
+                this.engine.bindFramebuffer(renderTo.getInternalTexture()!);
+            }
+            this.engine.bindBuffers(this._vertexBuffers, this._indexBuffer, effectWrapper.effect);
+            effectWrapper.onApplyObservable.notifyObservers({});
+
+            // Render
+            this.engine.drawElementsType(Constants.MATERIAL_TriangleFillMode, 0, 6);
+            if (renderTo) {
+                this.engine.unBindFramebuffer(renderTo.getInternalTexture()!);
+            }
+        });
+    }
+
+    /**
+     * Disposes of the effect renderer
+     */
+    dispose() {
+        if (this._tmpScreenBuffer) {
+            this._tmpScreenBuffer.dispose();
+        }
+    }
+}
+
+/**
+ * Wraps an effect to be used for rendering
+ */
+export class EffectWrapper {
+    private static _counter = 0;
+    /**
+     * Event that is fired right before the effect is drawn (should be used to update uniforms)
+     */
+    public onApplyObservable = new Observable<{}>();
+    /**
+     * The underlying effect
+     */
+    public effect: Effect;
+
+    /**
+     * Creates an effect to be renderer
+     * @param creationOptions options to create the effect
+     */
+    constructor(creationOptions: {engine: Engine, fragmentShader: string, attributeNames: Array<string>, uniformNames: Array<string>, samplerNames: Array<string>}) {
+        var shaderName = "EffectWrapperEffect" + EffectWrapper._counter++;

```

Comment:
I wonder why not using the same definition than effect where we can either pass the source or the shader store name.

It feels weird to touch the shader store if source code is provided. It could actually even be an element:

https://github.com/BabylonJS/Babylon.js/blob/master/src/Materials/effect.ts#L334